### PR TITLE
Load http_api module as a child process - Closes #3059

### LINKED ISSUE #3059: Load http_api module as a child process
### Expected behavior
As highlighted in the [LIP](https://github.com/LiskHQ/lips/blob/master/proposals/lip-0005.md#modules), implement the possibility of loading modules as child processes. 

Update module loading logic to load modules child process and establish RPC connection to the controller and load http_api module in a child process.
### Actual behavior
A module can be loaded and run on the same process as the controller.

### Which version(s) does this affect? (Environment, OS, etc...)
`1.6.0`

### PR DESCRIPTION
**Pull request by lsilvs**
### What was the problem?
Not possible to run modules in child process

### How did I fix it?
I've used nodejs `child_process` to fork new processes when the module is set to be loaded as child process.
The new `ChildProcessChannel` is used and communicate with the parent process through sockets.

### How to test it?
In `devnet` httpApi is loaded as child process by default. To change the configuration for other networks, set `ipc.enabled: true` in the config file.

### Review checklist

* The PR resolves #3059
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

### CODE REVIEW

File: framework/src/controller/controller.js [hunk 1]
```diff
@@ -203,23 +199,38 @@ class Controller {
...
-		channel.publish(`${moduleAlias}:loading:finished`);
+		const parameters = [
+			modulePath,
+			JSON.stringify({ ...options, socketsPath: this.socketsPath }),
+		];
+
+		const child = child_process.fork(program, parameters);
+
+		child.on('exit', () => {
+			this.logger.error(`Module ${moduleAlias}(${name}:${version}) exited`);
```
Comment (pablitovicente) [inline framework/src/controller/controller.js]: It would be nice here to have the child have an error handler that sends the error that cause the exit back to its parent so we have more information on the exit reason.

**Comment by yatki** [comment]
@lsilvs we need to have an environment variable to configure how we load modules. So we can execute functional tests with child process modules.

File: framework/src/controller/bus.js [hunk 2]
```diff
@@ -113,12 +118,12 @@ class Bus extends EventEmitter2 {
...
 		let channel;
-		if (options.type === 'inMemory') {
-			channel = options.channel;
-		} else {
+		if (options.rpcSocketPath) {
 			const rpcSocket = axon.socket('req');
 			rpcSocket.connect(options.rpcSocketPath);
 			channel = new RPCClient(rpcSocket);
+		} else {
+			channel = options.channel;
```
Comment (yatki) [inline framework/src/controller/bus.js]: I'd prefer to move this line to line 120 to remove else block.

File: framework/src/controller/bus.js [hunk 3]
```diff
@@ -178,17 +183,46 @@ class Bus extends EventEmitter2 {
...
 
-		super.emit(eventName, eventValue); // Communicate throw event emitter
-		this.pubSocket.emit(eventName, eventValue); // Communicate throw unix socket
+		// Communicate through event emitter
+		super.emit(eventName, eventValue);
+
+		// Communicate through unix socket
+		if (this.config.ipc.enabled) {
+			this.pubSocket.emit(eventName, eventValue);
+		}
```
Comment (yatki) [inline framework/src/controller/bus.js]: if ipc enabled, we are publishing 2 events here. What was the reason for that?

File: framework/src/controller/channels/child_process_channel.js [hunk 4]
```diff
@@ -29,33 +29,38 @@ class ChildProcessChannel extends BaseChannel {
...
+			this.rpcServer = new RPCServer(this.rpcSocket);
 
-		this.rpcSocket.bind(this.rpcSocketPath);
+			this.rpcServer.expose('invoke', (action, cb) => {
+				this.invoke(action)
+					.then(data => cb(null, data))
+					.catch(error => cb(error));
+			});
+
+			this.rpcSocket.bind(this.rpcSocketPath);
```
Comment (yatki) [inline framework/src/controller/channels/child_process_channel.js]: maybe we can move this line after 49 to keep the structure consistent?

File: framework/src/controller/application.js [hunk 5]
```diff
@@ -119,14 +119,20 @@ class Application {
 		__private.modules.set(this, {});
 		__private.transactions.set(this, {});
 
+		// TODO: move this configuration to module especific config file
+		const childProcessModules = process.env.CHILD_PROCESS_MODULES
```
Comment (shuse2) [inline framework/src/controller/application.js]: add `LISK_`?

File: framework/src/controller/bus.js [hunk 6]
```diff
@@ -38,40 +40,43 @@ class Bus extends EventEmitter2 {
...
-		this.rpcServer = new RPCServer(this.rpcSocket);
-		this.rpcSocket.bind(socketsPath.rpc);
-
-		this.rpcServer.expose(
-			'registerChannel',
-			(moduleAlias, events, actions, options, cb) => {
-				this.registerChannel(moduleAlias, events, actions, options)
-					.then(() => cb(null))
+	async setup() {
+		if (this.config.ipc.enabled) {
```
Comment (shuse2) [inline framework/src/controller/bus.js]: Prefer early return
```
if (!this.config.ipc.enabled) {
  return false;
}
```

File: framework/src/controller/bus.js [hunk 7]
```diff
@@ -178,17 +183,46 @@ class Bus extends EventEmitter2 {
...
 	subscribe(eventName, cb) {
 		if (!this.getEvents().includes(eventName)) {
-			throw new Error(`Event ${eventName} is not registered to bus.`);
+			this.logger.info(
+				`Event ${eventName} was subscribed but not registered to the bus yet.`
+			);
+		}
+
+		// Communicate through event emitter
+		super.on(eventName, cb);
```
Comment (shuse2) [inline framework/src/controller/bus.js]: `this.on`?

File: framework/src/controller/bus.js [hunk 8]
```diff
@@ -178,17 +183,46 @@ class Bus extends EventEmitter2 {
...
-		this.subSocket.on(eventName, cb); // Communicate throw unix socket
+	once(eventName, cb) {
+		if (!this.getEvents().includes(eventName)) {
+			this.logger.info(
+				`Event ${eventName} was subscribed but not registered to the bus yet.`
+			);
+		}
+
+		// Communicate through event emitter
+		super.once(eventName, cb);
```
Comment (shuse2) [inline framework/src/controller/bus.js]: `this.once`?

File: framework/src/controller/bus.js [hunk 9]
```diff
@@ -178,17 +183,46 @@ class Bus extends EventEmitter2 {
 			throw new Error(`Event ${eventName} is not registered to bus.`);
 		}
 
-		super.emit(eventName, eventValue); // Communicate throw event emitter
-		this.pubSocket.emit(eventName, eventValue); // Communicate throw unix socket
+		// Communicate through event emitter
+		super.emit(eventName, eventValue);
```
Comment (shuse2) [inline framework/src/controller/bus.js]: `this.emit`?

File: framework/src/controller/channels/child_process_channel.js [hunk 10]
```diff
@@ -142,44 +150,59 @@ class ChildProcessChannel extends BaseChannel {
...
-						}
-					);
-				});
-			}),
-		]);
+		const promises = [];
+
+		if (this.pubSocket) {
+			promises.push(
+				new Promise(resolve => {
```
Comment (yatki) [inline framework/src/controller/channels/child_process_channel.js]: for the sake of readability would be nice to have a function like below to promisify once events: 
```js
const promisifyOnce = (emitter, event, reject = false) => new Promise((resolve, reject) => {
  emitter.once(event, (...data) => {
    if (reject) {
      return reject(...data);
    }
    return resolve(...data);
  });
});

// Usage:

const subSocketConnectPromise = promisifyOnce(this.subSocket.sock, 'connect');
const subSocketConnectErrPromise = promisifyOnce(this.subSocket.sock, 'error', true);
```

File: framework/src/controller/channels/child_process_channel.js [hunk 11]
```diff
@@ -142,44 +150,59 @@ class ChildProcessChannel extends BaseChannel {
...
+				new Promise((resolve, reject) => {
+					this.busRpcSocket.once('connect', () => {
+						this.busRpcClient.call(
+							'registerChannel',
+							this.moduleAlias,
+							this.eventsList.map(event => event.name),
+							this.actionsList.map(action => action.name),
+							{ type: 'ipcSocket', rpcSocketPath: this.rpcSocketPath },
+							(err, result) => {
+								if (err) reject(err);
```
Comment (shuse2) [inline framework/src/controller/channels/child_process_channel.js]:
```
if (err) {
  reject(err);
}
```

File: framework/src/controller/child_process_loader.js [hunk 12]
```diff
@@ -0,0 +1,29 @@
+// Parameters passed by `child_process.fork(_, parameters)`
+const modulePath = process.argv[2];
+const { moduleOptions, config } = JSON.parse(process.argv[3]);
+
+// eslint-disable-next-line import/no-dynamic-require
+const Klass = require(modulePath);
+const { ChildProcessChannel } = require('./channels');
```
Comment (shuse2) [inline framework/src/controller/child_process_loader.js]: maybe static import first?

File: framework/src/controller/controller.js [hunk 13]
```diff
@@ -152,9 +150,18 @@ class Controller {
...
+				if (this.config.ipc.enabled) {
+					// eslint-disable-next-line no-await-in-loop
+					await this._loadChildProcessModule(alias, klass, options);
+				} else {
+					this.logger.warn(
+						`IPC is disabled. ${alias} will be loaded in-memory.`
+					);
+					// eslint-disable-next-line no-await-in-loop
+					await this._loadInMemoryModule(alias, klass, options);
+				}
```
Comment (yatki) [inline framework/src/controller/controller.js]: maybe if else statements can be simplified with using `continue;` ?

File: framework/src/controller/controller.js [hunk 14]
```diff
@@ -203,23 +213,42 @@ class Controller {
...
-			`Module ready with alias: ${moduleAlias}(${name}:${version})`
-		);
+		const child = child_process.fork(program, parameters);
+
+		child.on('exit', (code, signal) => {
+			this.logger.error(
+				`Module ${moduleAlias}(${name}:${version}) exited with code: ${code} and signal: ${signal}`
+			);
+			// TODO: Reload child instead of exiting the parent process
+			process.exit(1);
```
Comment (yatki) [inline framework/src/controller/controller.js]: Is this going to be fixed in another PR?

File: framework/test/jest/specs/unit/controller/channels/in_memory_channel.spec.js [hunk 15]
```diff
@@ -75,21 +73,21 @@ describe('InMemoryChannel Channel', () => {
 	describe('#constructor', () => {
 		it('should create the instance with given arguments.', () => {
 			// Assert
-			expect(inMemoryChannel.bus).toBe(params.bus);
+			expect(inMemoryChannel.bus).toBe(undefined);
```
Comment (yatki) [inline framework/test/jest/specs/unit/controller/channels/in_memory_channel.spec.js]: this can be replaced with `toHaveProperty` matcher.

File: framework/src/controller/bus.js (hunk 7 above)
Comment (yatki) [reply framework/src/controller/bus.js]: I kind of prefer to use `super` here because it's instantly clear that we are using a parent class method. But I don't have strong opinion on this.

File: framework/src/controller/bus.js (hunk 3 above)
Comment (lsilvs) [reply framework/src/controller/bus.js]: First emit is for in-memory modules and the second one for child process

File: framework/src/controller/controller.js (hunk 13 above)
Comment (lsilvs) [reply framework/src/controller/controller.js]: We have a eslint rule `no-continue`. That's why I didn't use it.
We shouldn't be using `for of` with `await` anyway.
Next iteration we can improve this.

File: framework/src/controller/controller.js (hunk 14 above)
Comment (lsilvs) [reply framework/src/controller/controller.js]: https://github.com/LiskHQ/lisk/issues/3142

File: framework/test/jest/specs/unit/controller/channels/in_memory_channel.spec.js (hunk 15 above)
Comment (lsilvs) [reply framework/test/jest/specs/unit/controller/channels/in_memory_channel.spec.js]: Can you give an example, pls? I've changed to `undefined` because we not setting the bus in the constructor anymore.

File: framework/src/controller/bus.js (hunk 8 above)
Comment (lsilvs) [reply framework/src/controller/bus.js]: Here we can't use `this` because `this.once` is its own method and we get `Maximum call stack size exceeded`

File: framework/src/controller/channels/child_process_channel.js (hunk 10 above)
Comment (lsilvs) [reply framework/src/controller/channels/child_process_channel.js]: This code was part of another PR and I prefer keeping it as it is in this PR.

File: framework/src/controller/bus.js (hunk 8 above)
Comment (shuse2) [reply framework/src/controller/bus.js]: ah my bad, this will be recursive... =p

File: framework/src/controller/controller.js (hunk 13 above)
Comment (shuse2) [reply framework/src/controller/controller.js]: > We shouldn't be using for of with await anyway.
Not related to this but what's the reason for this? i think this is the only way though

Comment (lsilvs) [reply framework/src/controller/controller.js]: afaik `for await...of` was created for this

File: framework/test/jest/specs/unit/controller/channels/in_memory_channel.spec.js (hunk 15 above)
Comment (yatki) [reply framework/test/jest/specs/unit/controller/channels/in_memory_channel.spec.js]: Ah in that case please move this test from constructor to registerToBus suite.

Also just fyi I meant `expect(inMemoryChannel).toHaveProperty('bus')` could be used if the property is declared but was not assigned yet.

**Comment by lsilvs** [comment]
@diego-G Can you investigate your local environment better, pls?
It works perfectly on Jenkins and all tests are green.
